src/mds/Server.cc

```diff
@@ -2552,7 +2552,39 @@ void Server::dispatch_client_request(MDRequestRef& mdr)
     respond_to_request(mdr, mdr->more()->peer_error);
     return;
   }
-  
+
+  Session *session = mds->get_session(req);
+  if (session && !session->info.has_feature(CEPHFS_FEATURE_ALTERNATE_NAME) &&
+      req->get_op() != CEPH_MDS_OP_LOOKUP &&
+      req->get_op() != CEPH_MDS_OP_LOOKUPHASH &&
+      req->get_op() != CEPH_MDS_OP_LOOKUPPARENT &&
+      req->get_op() != CEPH_MDS_OP_LOOKUPINO &&
+      req->get_op() != CEPH_MDS_OP_LOOKUPNAME &&
+      req->get_op() != CEPH_MDS_OP_LOOKUPSNAP &&
+      req->get_op() != CEPH_MDS_OP_RMSNAP &&
+      req->get_op() != CEPH_MDS_OP_LSSNAP &&
+      req->get_op() != CEPH_MDS_OP_GETATTR &&
+      req->get_op() != CEPH_MDS_OP_READDIR &&
+      req->get_op() != CEPH_MDS_OP_UNLINK &&
+      req->get_op() != CEPH_MDS_OP_RMDIR) {
+    CInode *cur = mdcache->get_inode(req->get_filepath().get_ino());
+    if (!cur)
+      return;
+    MutationImpl::LockOpVec lov;
+    /* We need 'As' caps for the fscrypt context */
+    lov.add_rdlock(&cur->authlock);
```

Comment:
This is causing a deal loop between `setattr` and `cap revoke`:

https://pulpito.ceph.com/vshankar-2023-07-25_11:29:34-fs-wip-vshankar-testing-20230725.043804-testing-default-smithi/7350758/

If there has only a signle old client, then this client could get the `Asx` caps always.

1, When truncating the file, it will try to acquire the `rdlock` for `authlock`. And then mds will try to revoke the `Ax` caps from the old client. And then this truncating request goes to sleep. Since the old client is the only client the `Fsxrwcb` caps will be granted to this old client.

2, Client releases the `Ax` caps. Then revoke the above truncating request.

3, This time it successfully gets that the `fscrypt` is not enabled and then continue to call `handle_client_setattr()`. But this time it will try to `xlock` the `filelock` and then need to revoke the `Fsrxw` caps from old client and then goes to sleep. But since there has only one client the `Asx` caps will be granted to the old client with the `revoke` request.

4, client released the `Fsrxw` caps and then wakes up the truncating request, and then goes to `step 1`.

Then it goes to a dead loop until the qa test timedout.

Reply:
@luis-henrix Could you check this ? thanks

Reply:
To break the dead loop we should do this check in each OPs instead of here just after acquired all the needed locks.